In [109]:
import os
import pandas as pd

import plotly.express as px
import plotly.offline as pyo

pyo.init_notebook_mode()

Output hidden; open in https://colab.research.google.com to view.

### Importing Dataset

In [110]:
from google.colab import drive 
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [111]:
users_data = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Tegger/users_data.csv')
snowplow_data = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Tegger/snowplow_data.csv')

In [112]:
print(users_data.shape)
print(snowplow_data.shape)

(3764, 15)
(23465, 51)


### Understanding Dataset

In [113]:
users_data.head(5)

Unnamed: 0  ... engagement_time_minutes
0           0  ...                    2.42
1           1  ...                    2.00
2           2  ...                    1.89
3           3  ...                    1.38
4           4  ...                    0.47

[5 rows x 15 columns]

In [114]:
users_data.dtypes

Unnamed: 0                   int64
user_id                     object
snowplow_pseudo_id          object
AGEB_key                    object
user_first_visit            object
traffic_source              object
device_category             object
browser                     object
mobile_brand                object
model_name                  object
country                     object
total_session_duration       int64
total_events                 int64
total_page_views             int64
engagement_time_minutes    float64
dtype: object

In [115]:
snowplow_data.head(5)

Unnamed: 0  altitude  ...                               user_id  pseudo_user_id
0           0       NaN  ...  4888686b-3d8e-4baf-82c1-c2ba0faffa62    1.890160e+08
1           1       NaN  ...  4888686b-3d8e-4baf-82c1-c2ba0faffa62    1.890160e+08
2           2       NaN  ...  4888686b-3d8e-4baf-82c1-c2ba0faffa62    1.890160e+08
3           3       NaN  ...  b15477c0-41cc-483c-9e3c-4a439ce2df34    3.148460e+08
4           4       NaN  ...  4888686b-3d8e-4baf-82c1-c2ba0faffa62    1.890160e+08

[5 rows x 51 columns]

In [116]:
snowplow_data.dtypes

Unnamed: 0                     int64
altitude                     float64
altitudeAccuracy             float64
bearing                      float64
br_colordepth                  int64
br_family                     object
br_lang                       object
br_name                       object
br_type                       object
br_viewheight                  int64
br_viewwidth                   int64
derived_tstamp                object
derived_tstamp_local          object
doc_height                     int64
doc_width                      int64
domain_sessionid              object
domain_userid                 object
dvce_ismobile                  int64
dvce_screenheight              int64
dvce_screenwidth               int64
dvce_type                     object
event_id                      object
event_name                    object
geo_city                      object
geo_country                   object
geo_latitude                 float64
geo_longitude                float64
g

In [117]:
## chosse rows which matches the users_id in the list

# users_data.query('user_id in @users_id')
# users_data[users_data['user_id'].isin(users_id)]


# creating a list of User IDs

users_id =users_data['user_id']

print("Before picking out unique values")
print(users_id.shape)
print(users_id.value_counts())

print("\nAfter picking our unique values")
users_id = users_id.unique()
print(users_id.shape)

Before picking out unique values
(3764,)
2be3c4b4-081c-4721-9469-ae2d0a6af473    544
be9335e4-64e4-483b-85db-9eda502802fc    132
91e978ea-78e6-4bfe-96db-2d8b56f87c78    109
1ad495ad-2e1b-4a37-ae3a-9f1dd8fe2931     96
1c101e61-f3cc-4d77-8803-e27aa80f2a59     96
                                       ... 
18c05087-2f95-4639-b354-4b896cb55e7a      1
831edb2a-f16c-4414-9cef-10876f0fc339      1
80267c71-54e1-4e1f-8035-b1d064cee25e      1
fd71592b-47d9-4e0c-908a-79a947162634      1
d93275fb-097f-4d65-ae18-de7321ca9d82      1
Name: user_id, Length: 859, dtype: int64

After picking our unique values
(859,)


There are 859 unique ids who have total sessions time

In [118]:
snowplow_users_id =snowplow_data['user_id'].unique()
print(snowplow_users_id.shape)

(32,)


There are only 32 unique ids in snowplow_data. This mean only for 32 users in users_data we have addtional data in snowplow_data.

## Q1. Is there a correlation between screen size and session duration?

### Determining the measure of screen size: 'device_category' or 'dvce_screenheight', and 'dvce_screenwidth'.

In [119]:
# for random user 1
print(users_data[users_data['user_id']=='2be3c4b4-081c-4721-9469-ae2d0a6af473']['device_category'].value_counts())
print(snowplow_data[snowplow_data['user_id']=='2be3c4b4-081c-4721-9469-ae2d0a6af473']['dvce_screenheight'].value_counts())

desktop    425
mobile      78
tablet      41
Name: device_category, dtype: int64
480    1993
540     555
Name: dvce_screenheight, dtype: int64


In [120]:
# for random user 2
print(users_data[users_data['user_id']=='1c101e61-f3cc-4d77-8803-e27aa80f2a59']['device_category'].value_counts())
print(snowplow_data[snowplow_data['user_id']=='c101e61-f3cc-4d77-8803-e27aa80f2a59']['dvce_screenheight'].value_counts())

desktop    96
Name: device_category, dtype: int64
Series([], Name: dvce_screenheight, dtype: int64)


Not all user ids have device screen height and screen width. So, taking device category seems a better choice for measuring screen size

### Finding whether there exist correlation between 'device_category' and 'total_session_duration'


In [121]:
df1 = users_data.filter(['user_id','device_category','total_session_duration'], axis=1)
print(df1.shape)

(3764, 3)


In [122]:
df1.head(5)

user_id device_category  total_session_duration
0  e3eabd43-23f3-4781-90ee-5a67e781e7e2         desktop                    2019
1  e4ccc0ac-71b1-4d7c-b697-f9f76f1084d0          mobile                     467
2  815ef036-f1ee-486e-85a2-facf11df24a1          mobile                     540
3  c86127fb-bc84-488a-b961-2abc5213f6fd          mobile                     161
4  1c101e61-f3cc-4d77-8803-e27aa80f2a59         desktop                      81

In [123]:
fig = px.box(df1, x ='device_category', y = 'total_session_duration')
fig.show(renderer="colab")

#### Carrying out ANOVA testing to determine existence of correlation or not?

In [124]:
grps = pd.unique(df1.device_category)
print(grps)
data_groups = {grp:df1['total_session_duration'][df1['device_category']==grp] for grp in grps}

['desktop' 'mobile' 'tablet']


In [125]:
k = len(grps)                                # number of independent comaparison groups
print(k)
N = len(df1.values)                          # total number of samples
print(N)

# n = df1.groupby('device_category')['total_session_duration'].nunique()
n = dict(df1.groupby('device_category').size()) # number of samples in each group
print(n)

3
3764
{'desktop': 2255, 'mobile': 1430, 'tablet': 79}


In [126]:
from scipy import stats

F, p = stats.f_oneway(data_groups['desktop'], data_groups['mobile'], data_groups['tablet'])

F, p

(1.422647570398177, 0.24120460879072453)

Ans: Here, we have a p-value of 0.24 which is greater than 0.05 and 0.1 at 5 % and 10 % significance level. This means that the our sample statistics F isn't statistically significant to reject the nulle hypothesis and there is no correlation between screen size and total session duration both at 5% and 10 % significance level. 

## Q2. This data includes three partner sites: Freim.tv, News culturacolectiva.com, and culturacolcetiva.com. Based on Page_URL on which of the three partner sites do users have longer sessions?

In [127]:
df2 = users_data.filter(['user_id', 'traffic_source', 'total_session_duration'], axis=1)
print(df2.shape)

(3764, 3)


In [128]:
df2.head(5)

user_id  traffic_source  total_session_duration
0  e3eabd43-23f3-4781-90ee-5a67e781e7e2        (direct)                    2019
1  e4ccc0ac-71b1-4d7c-b697-f9f76f1084d0  m.facebook.com                     467
2  815ef036-f1ee-486e-85a2-facf11df24a1          google                     540
3  c86127fb-bc84-488a-b961-2abc5213f6fd  m.facebook.com                     161
4  1c101e61-f3cc-4d77-8803-e27aa80f2a59       tegger.io                      81

In [129]:
df2['traffic_source'].value_counts()

tegger.io                                       1361
(direct)                                        1222
google                                           405
culturacolectiva.com                             251
news.culturacolectiva.com                        183
freim.tv                                         145
m.facebook.com                                    36
l.facebook.com                                    30
bing                                              21
t.co                                              20
culturacolectiva-com.cdn.ampproject.org           17
Cultura Colectiva                                 12
pinterest.com                                      8
hora724.com                                        8
taboola                                            6
blog.propsproject.com                              4
newsletter                                         3
lm.facebook.com                                    2
hu.pinterest.com                              

In [130]:
cat_friem = df2[df2['traffic_source'].isin(['freim.tv'])]
print(cat_friem.shape)
print(sum(cat_friem['total_session_duration']))

(145, 3)
12663667


In [131]:
cat_cult = df2[df2['traffic_source'].isin(['culturacolectiva.com','culturacolectiva-com.cdn.ampproject.org', 'Cultura Colectiva' ])]
print(cat_cult.shape)
print(sum(cat_cult['total_session_duration']))

(280, 3)
30335473


In [132]:
cat_news_cult = df2[df2['traffic_source'].isin(['news-culturacolectiva-com.cdn.ampproject.org','news.culturacolectiva.com'])]
print(cat_news_cult.shape)
print(sum(cat_news_cult['total_session_duration']))

(184, 3)
57543840


#### Ans: News.culturalcolective.com has the longer sessions among the three partner sites

## Q3. What combination of traffic source, browser language, browser, and mobile brand yields users with the longest sessions? What combination yields users with more page views?

In [134]:
# we already have traffic_source, browser, mobile brand, total session duration, total page views in user_data
# Let's create a new dataframe df3 with empyty column br_language which we need to fill from snowplow_data

df3 = users_data.filter(['user_id','traffic_source', 'browser','mobile_brand','total_session_duration', 'total_page_views'],axis=1)
df3['br_lang']= ""
print(df3.shape)

(3764, 7)


In [136]:
# for each user_id get the browser language from snowplot_data and put it into the br_lang column in df3
# It was confirmed that almost all the users browse the sites in a single language

for i in range(len(users_id)):
  temp_array = snowplow_data[snowplow_data['user_id']==users_id[i]]['br_lang'].value_counts().index.tolist()
  if temp_array:
    lang = temp_array[0]
  else:
    lang = ""
  df3.loc[df3['user_id'].isin([users_id[i]]), 'br_lang'] = str(lang)

df3['br_lang'].value_counts()

          2730
es-ES      544
en-US      353
de-DE       66
ko          38
es-mx       15
es-MX       13
es           2
es-AR        1
es-419       1
es-US        1
Name: br_lang, dtype: int64

For 70 % of the data in user_data, there was no br_lang found in snowplow_data. Hence, for such data an empty string was put in df3.

In [137]:
df3.head(5)

user_id  ... br_lang
0  e3eabd43-23f3-4781-90ee-5a67e781e7e2  ...        
1  e4ccc0ac-71b1-4d7c-b697-f9f76f1084d0  ...        
2  815ef036-f1ee-486e-85a2-facf11df24a1  ...        
3  c86127fb-bc84-488a-b961-2abc5213f6fd  ...        
4  1c101e61-f3cc-4d77-8803-e27aa80f2a59  ...   en-US

[5 rows x 7 columns]

In [139]:
traffic_cat = df3['traffic_source'].value_counts().index.to_list()
print("Num of Traffic Source: %d " % (len(traffic_cat)))
browser_cat = df3['browser'].value_counts().index.to_list()
print("Num of Browser: %d " % (len(browser_cat)))
mobile_cat = df3['mobile_brand'].value_counts().index.to_list()
print("Num of Mobile brands: %d" % (len(mobile_cat)))
browser_lang_cat = df3['br_lang'].value_counts().index.tolist()
print("Num of Browser Languages: %d" % (len(browser_lang_cat)))

Num of Traffic Source: 42 
Num of Browser: 11 
Num of Mobile brands: 31
Num of Browser Languages: 11


So there are 157542 possible combinations.

In [140]:
print(traffic_cat)
print(browser_cat)
print(mobile_cat)
print(browser_lang_cat)

['tegger.io', '(direct)', 'google', 'culturacolectiva.com', 'news.culturacolectiva.com', 'freim.tv', 'm.facebook.com', 'l.facebook.com', 'bing', 't.co', 'culturacolectiva-com.cdn.ampproject.org', 'Cultura Colectiva', 'pinterest.com', 'hora724.com', 'taboola', 'blog.propsproject.com', 'newsletter', 'lm.facebook.com', 'hu.pinterest.com', 'ampproject.org', 'news-staging.culturacolectiva.com', 'es.search.yahoo.com', 'news.google.com', 'mymoney777.icu', 'msn.com', 'pinterest.ca', 'ecosia.org', 'mx.search.yahoo.com', 'tejidosacrochetpasoapaso.com', 'nattittorres.wixsite.com', 'l.instagram.com', 'yahoo', 'propeller', 'pinterest.com.mx', 'aulavirtualinmaculadaloja.edu.ec', 'markets.businessinsider.com', 'ejemplos.co', 'id.pinterest.com', 'facebook.com', 'news-culturacolectiva-com.cdn.ampproject.org', 'transparentbusiness.com', 'duckduckgo']
['Chrome', 'Safari', 'Firefox', 'Edge', 'Opera', 'Samsung Internet', 'Android Webview', 'Puffin', 'YaBrowser', 'Safari (in-app)', 'UC Browser']
['Google', 

In [ ]:
rows_list = []
for i in traffic_cat:
  for j in browser_cat:
    for k in browser_lang_cat:
      for l in mobile_cat:
        result = df3[(df3['traffic_source']==i) & (df3['browser']==j) & (df3['br_lang']==k) & (df3['mobile_brand']==l)]
        if not result.empty:
          total_ses = sum(result['total_session_duration'])
          total_page = sum(result['total_page_views'])
          dict = {}
          dict.update({'traffic_source':i, 'browser': j, 'br_lang':k, 'mobile_brand': l ,'total_session_duration':total_ses, 'total_page_views':total_page})
          rows_list.append(dict)

df3_combination = pd.DataFrame(rows_list)


In [ ]:
print(df3_combination.shape)
df3_combination.head(10)

(217, 6)


traffic_source browser  ... total_session_duration total_page_views
0      tegger.io  Chrome  ...               44144569           385096
1      tegger.io  Chrome  ...                 205306            13689
2      tegger.io  Chrome  ...               62461943         23182122
3      tegger.io  Chrome  ...                1070659            28858
4      tegger.io  Chrome  ...                5555936            17122
5      tegger.io  Chrome  ...                1390944            33515
6      tegger.io  Chrome  ...                  55404              328
7      tegger.io  Chrome  ...                  13524               90
8      tegger.io  Chrome  ...                  90970              603
9      tegger.io  Chrome  ...                 143441             1607

[10 rows x 6 columns]

In [ ]:
df3_combination[df3_combination['total_session_duration']==df3_combination['total_session_duration'].max()]

traffic_source browser  ... total_session_duration total_page_views
2      tegger.io  Chrome  ...               62461943         23182122

[1 rows x 6 columns]

Combination of traffic source, browser language, browser, and mobile brand that has longest total session time shown above.

In [ ]:
df3_combination[df3_combination['total_page_views']==df3_combination['total_page_views'].max()]

traffic_source browser  ... total_session_duration total_page_views
2      tegger.io  Chrome  ...               62461943         23182122

[1 rows x 6 columns]

Combination of traffic source, browser language, browser, and mobile brand that has most page views shown above.